### PR TITLE
Support inferring&completing `Self` type in enum/struct/union definitions

### LINKED ISSUE #1908: Does not infer `Self`
image

```rust
enum Test {
    Me(Box<Self>),
}

fn test(t: Test) {
    match t {
        Test::Me(bug) => (),
    }
}
```

Type of `bug` is not inferred.
Rivision: d9b4835625ac96c7628e2ef66ef6e26cc48d231f

### PR DESCRIPTION
**Pull request by ice1000**
Signed-off-by: ice1000 <ice1000kotlin@foxmail.com>

An attempt to fix #1908.
This code works, but I believe the implementation is ugly. Please give me suggestions!

### CODE REVIEW

**Comment by ice1000** [comment]
I've updated the comment in #1908 with a screenshot.

**Comment by flodiebold** [comment]
Note that this should also work for structs; also, `Self` should refer to the whole type with its type parameters, not just the enum. E.g.:
```rust
struct List<T> {
    elem: T,
    next: Option<Box<Self>>,
}
```
(For a test, I'd shorten that to `struct List<T> { next: Self }` -- it doesn't matter that this wouldn't compile.)

We already resolve `Self` in impl blocks, and this is very similar. The `Resolver` works by having a stack of scopes of various kinds, one of them [`ImplBlock`](https://github.com/rust-analyzer/rust-analyzer/blob/21fa889cf3f70c507e3b9f2f6362e65cbb8ed955/crates/ra_hir/src/resolve.rs#L47). When building a `Resolver` for an impl block, we push such a scope on the stack [here](https://github.com/rust-analyzer/rust-analyzer/blob/21fa889cf3f70c507e3b9f2f6362e65cbb8ed955/crates/ra_hir/src/impl_block.rs#L106). The `ImplBlockScope` is then handled in a few places in `resolve.rs`; when resolving `Self` as a type, we return [`TypeNs::SelfType`](https://github.com/rust-analyzer/rust-analyzer/blob/21fa889cf3f70c507e3b9f2f6362e65cbb8ed955/crates/ra_hir/src/resolve.rs#L54). And that's then mainly handled [here](https://github.com/rust-analyzer/rust-analyzer/blob/21fa889cf3f70c507e3b9f2f6362e65cbb8ed955/crates/ra_hir/src/ty/lower.rs#L177) during type lowering to resolve the target type of the impl.

I think I'd suggest adding an `AdtScope(Adt)` in the `Scope` enum, and an `AdtSelfType(Adt)` in the `TypeNs` enum. (Note that this is different from the existing `TypeNs::Adt`, because that one takes type parameters, while `Self` always gets the type parameters from the definition.) The scope should be handled basically the same way as the impl block scope; we just need to take care to push it in the `resolver` methods of `Struct`, `Enum` and `Union`. And when handling `TypeNs::AdtSelfType` in `lower.rs`, you can just use the `ty` method to get the definition type.

(Handling the scope [here](https://github.com/rust-analyzer/rust-analyzer/blob/21fa889cf3f70c507e3b9f2f6362e65cbb8ed955/crates/ra_hir/src/resolve.rs#L439) will also mean we'll complete `Self` in struct/enum definitions, by the way.)

**Comment by ice1000** [comment]
Changed implementation into `Scope::AdtSelfScope`, please take a look @flodiebold

File: crates/ra_hir/src/code_model.rs [hunk 1]
```diff
@@ -465,7 +465,7 @@ impl Enum {
         // ...and add generic params, if present
         let p = self.generic_params(db);
         let r = if !p.params.is_empty() { r.push_generic_params_scope(p) } else { r };
-        r
+        r.push_scope(Scope::AdtScope(From::from(self)))
```
Comment (matklad) [inline crates/ra_hir/src/code_model.rs]:
```suggestion
        r.push_scope(Scope::AdtScope(self.into))
```

File: crates/ra_hir/src/resolve.rs [hunk 2]
```diff
@@ -43,8 +43,10 @@ pub(crate) enum Scope {
     ModuleScope(ModuleItemMap),
     /// Brings the generic parameters of an item into scope
     GenericParams(Arc<GenericParams>),
-    /// Brings `Self` into scope
+    /// Brings `Self` in `impl` block into scope
     ImplBlockScope(ImplBlock),
+    /// Brings `Self` in enum definition into scope
```
Comment (matklad) [inline crates/ra_hir/src/resolve.rs]:
```suggestion
    /// Brings `Self` in enum, struct and union definitions into scope
```

File: crates/ra_hir/src/ty/infer.rs [hunk 3]
```diff
@@ -559,6 +559,9 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             match resolver.resolve_path_in_type_ns_fully(self.db, &path) {
                 Some(TypeNs::Adt(Adt::Struct(it))) => it.into(),
                 Some(TypeNs::Adt(Adt::Union(it))) => it.into(),
+                Some(TypeNs::AdtSelfType(Adt::Struct(it))) => it.into(),
+                Some(TypeNs::AdtSelfType(Adt::Union(it))) => it.into(),
+                Some(TypeNs::AdtSelfType(Adt::Enum(it))) => it.into(),
```
Comment (matklad) [inline crates/ra_hir/src/ty/infer.rs]: I think these three lines could be replaced with 

```rust
Some(TypeNs::AdtSelfType(adt)) => adt.into()
```

**Comment by flodiebold** [comment]
Looks good to me apart from what @matklad already wrote! You _could_ add a test for completing `Self` in this situation (see [here](https://github.com/rust-analyzer/rust-analyzer/blob/93199002af05f3a3dfd274fe10633372f2471b73/crates/ra_ide_api/src/completion/complete_scope.rs#L471)).

Feel free to r+ with @matklad's suggestions:

bors delegate+

**Comment by ice1000** [comment]
I'm gonna apply the review suggestions locally instead of using GitHub.

**Comment by ice1000** [comment]
I'm very nervous. It's my first time to summon bors for merging a PR.

**Comment by matklad** [comment]
@ice1000 we have a rather extensive test suite, so the probability that something breaks spectacularly is pretty low :) Just reply with bors r+ comment

**Comment by ice1000** [comment]
Between the darkness and the dawn, a PR bot rises!

bors r+

**Pull request merged**